call_option.go

```diff
@@ -1,118 +1,116 @@
+// Copyright 2016, Google Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 package gax
 
 import (
+	"math/rand"
 	"time"
 
+	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 )
 
 type CallOption interface {
 	Resolve(*CallSettings)
 }
 
-type callOptions []CallOption
+type retryerOption func() Retryer
+
+func (o retryerOption) Resolve(s *CallSettings) {
+	s.Retryer = o
+}
 
-func (opts callOptions) Resolve(s *CallSettings) *CallSettings {
-	for _, opt := range opts {
-		opt.Resolve(s)
+func WithBackoffRetryer(base, max time.Duration, mult float64, codes ...codes.Code) CallOption {
```

Comment:
I think you are slightly misunderstanding the intention of Base/Initial and Max. I think the intention is best illustrated with an example. If we have
```
BackoffRetryer{
  Initial: time.Second,
  Max: 10*time.Second,
  Mult: 2,
  Codes: ...
}
```

The sleep times will be
- random value between 0s and 1s
- random value between 0s and 2s
- random value between 0s and 4s
- random value between 0s and 8s
- random value between 0s and 10s
- random value between 0s and 10s
- ...

and keeps going on until `ctx.Done()` closes. I believe this was proposed by @okdave with the rationale that most of our clients will be web servers. The contexts they use to connect to our services can just be the same as the ones they use to handle their requests. Then, the client will just keep trying until their user gives up.

The default values are not important to VKit, since VKit always fills all values. For use by non-VKit code though, I think Max should default to something like 30s.

Please let me know what you think.